x/outline/outline_device.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package outline
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"strconv"
+	"sync"
+
+	"github.com/Jigsaw-Code/outline-sdk/network"
+	"github.com/Jigsaw-Code/outline-sdk/network/dnstruncate"
+	"github.com/Jigsaw-Code/outline-sdk/network/lwip2transport"
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+	"github.com/Jigsaw-Code/outline-sdk/transport/shadowsocks"
+	"github.com/Jigsaw-Code/outline-sdk/x/connectivity"
+)
+
+const (
+	connectivityTestDomain   = "www.google.com"
+	connectivityTestResolver = "1.1.1.1:53"
+)
+
+type OutlineConfig struct {
+	Hostname string
+	Port     uint16
+	Password string
+	Cipher   string
+}
+
+type OutlineDevice struct {
+	t2s              network.IPDevice
+	pktProxy         network.DelegatePacketProxy
+	fallbackPktProxy network.PacketProxy
+	ssStreamDialer   transport.StreamDialer
+	ssPktListener    transport.PacketListener
+	ssPktProxy       network.PacketProxy
+}
+
+func NewOutlineDevice(accessKey string) (*OutlineDevice, error) {
```

Comment:
This conflates too many things to be in the SDK. It's code that belongs in outline-client instead.

For the SDK, I'd like to see a device others can reuse that is detached from our application logic. And users should be able to provide their own relay logic. We can provide the Outline and relay functionality via composition/layering.

One functionality that any VPN provider will need is the UDP refresh.
It would be helpful to see an IPdevice that provides the UDP update, taking the transport objects.

The relay should be a separate function. So is the Outline/Shadowsocks logic.

Reply:
removed this and replaced with three new functions: `NewOutlineStreamDialer`, `NewOutlinePacketProxy` and `NewOutlinePacketListener`.